Tests/MongoSwiftSyncTests/SpecTestRunner/TestOperationResult.swift

```diff
@@ -120,3 +154,128 @@ extension DeleteResult: BulkWriteResultConvertible {
         BulkWriteResult.new(deletedCount: self.deletedCount)
     }
 }
+
+struct ErrorResult: Equatable, Matchable {
+    internal static let errorKeys: Set = ["errorContains", "errorCodeName", "errorLabelsContain", "errorLabelsOmit"]
+
+    internal var errorContains: String?
+
+    internal var errorCodeName: String?
+
+    internal var errorLabelsContain: [String]?
+
+    internal var errorLabelsOmit: [String]?
+
+    public init(from doc: Document) {
+        let errorLabelsContain = doc["errorLabelsContain"]?.arrayValue?.compactMap { $0.stringValue }
+        let errorLabelsOmit = doc["errorLabelsOmit"]?.arrayValue?.compactMap { $0.stringValue }
+
+        self.errorContains = doc["errorContains"]?.stringValue
+        self.errorCodeName = doc["errorCodeName"]?.stringValue
+        self.errorLabelsContain = errorLabelsContain?.sorted()
+        self.errorLabelsOmit = errorLabelsOmit?.sorted()
+    }
+
+    internal static func == (lhs: ErrorResult, rhs: ErrorResult) -> Bool {
+        lhs.errorContains == rhs.errorContains &&
+            lhs.errorCodeName == rhs.errorCodeName &&
+            lhs.errorLabelsContain == rhs.errorLabelsContain &&
+            lhs.errorLabelsOmit == rhs.errorLabelsOmit
+    }
+
+    public func checkErrorResult(_ error: Error) throws {
```

Comment:
so we still have the problem where we're deriving our assertions from the type of the error the driver threw rather than the expectations from the test. So instead of casting the error first and checking the test requirements, we need to first check for the test requirements and cast as necessary to verify.

e.g.
```swift
if let messageSubstring = self.errorContains {
    expect(error.localizedDescription).to(contain(messageSubstring))
}

if let codeName = self.errorCodeName {
    if let commandError = error as? CommandError {
        expect(commandError.codeName).to(equal(codeName))
    } else if let ... = error as? ... { ... }
    else { XCTFail("\(error) doesn't contain a codeName") }
}

if let errorLabels = self.errorLabelsContain {
    guard let labeledError = error as? LabeledError else {
        XCTFail("\(error) doesn't contain errorLabels")
    }
    expect(labeledError.errorLabels).to(contain(errorLabels))
}

// etc
```
This is important because as it is now, a failing test could mistakeln pass depending on the error thrown. For example, if the test specifies `errorLabelsContain: ["some label"]` and the driver threw a `LogicError`, we would pass because `checkLogicError` doesn't verify errorLabels.

Reply:
Ahh apologies for the confusion, I've made the proper changes. One thing of note is that `localizedDescription`doesn't return the correct error message so I've had to resort to parsing the error message depending on the error type. Another thing of note is that the messages have case discrepancies so I've compared them in lowercase.

Reply:
Sounds good. I think we probably want our `localizedDescription` to contain a `message`, if there is one. Filed [SWIFT-769](https://jira.mongodb.org/browse/SWIFT-769) to fix that.